### PR TITLE
Implement torch.squareform

### PR DESCRIPTION
**Pull request by lhuang04**
Summary:
implement the equivalent of
https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.squareform.html#scipy.spatial.distance.squareform
as torch.squareform.

Differential Revision: D13568360

Test Plan:
# Basic functional test
python test/test_torch.py TestTorch.test_squareform 
# Autograd
python test/test_autograd.py  TestAutograd.test_squareform_1d TestAutograd.test_squareform_2d TestAutograd.test_squareform_1d_large TestAutograd.test_squareform_2d_large

# Benchmark in ipython
import torch
from scipy.spatial.distance import squareform, pdist
cuda = torch.device('cuda')
#1024x1024
v523776 = torch.arange(1, 523777, dtype=torch.double)
#2048x2048
v2096128 = torch.arange(1, 2096129, dtype=torch.double)
#10000x10000
v49995000 = torch.arange(1, 49995001, dtype=torch.double)

c523776 = torch.arange(1, 523777, dtype=torch.double, device = cuda)
c2096128 = torch.arange(1, 2096129, dtype=torch.double, device = cuda)
c49995000 = torch.arange(1, 49995001, dtype=torch.double, device = cuda)
lst = [v523776, v2096128, v49995000]
clst = [c523776, c2096128, c49995000]

def foo(lst, f):
  for vec in lst:
    ysq = f(vec)
    yvec = f(ysq)

timeit foo(lst, squareform)

timeit foo(lst, torch.squareform)

timeit foo(clst, torch.squareform)

# scipy
In [5]: timeit foo(lst, squareform)
3.62 s ± 28.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
# torch cpu
In [6]: timeit foo(lst, torch.squareform)
2.11 s ± 49.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
# torch cuda
In [7]: timeit foo(clst, torch.squareform)
41.6 ms ± 11.3 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

### CODE REVIEW

File: torch/nn/functional.py [hunk 1]
```diff
@@ -2783,6 +2783,36 @@ def pairwise_distance(x1, x2, p=2., eps=1e-6, keepdim=False):
...
+    array([[ 0,  1,  2,  3,  4],
+           [ 1,  0,  5,  6,  7],
+           [ 2,  5,  0,  8,  9],
+           [ 3,  6,  8,  0, 10],
+           [ 4,  7,  9, 10,  0]])
+    >>> F.squareform(sqm)
+    tensor([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10])
+
+    """
+    assert input.dim() == 1 or input.dim() == 2, 'dimension of the input need be 1 or 2'
```
Comment (gchanan) [inline torch/nn/functional.py]: squareform should be a function at the C++ (native_function) level.  The downside of doing it this way is that the C++ API does not match the python API.

File: aten/src/ATen/native/native_functions.yaml [hunk 2]
```diff
@@ -1261,6 +1261,18 @@
 - func: cosine_similarity(Tensor x1, Tensor x2, int64_t dim=1, double eps=1e-8) -> Tensor
   variants: function
 
+- func: to_squareform(Tensor self) -> Tensor
+  python_module: nn
```
Comment (gchanan) [inline aten/src/ATen/native/native_functions.yaml]: these shouldn't go int the nn module, this function doesn't really have anything to do with neural networks specifically.

File: aten/src/ATen/native/Distance.cpp [hunk 3]
```diff
@@ -75,4 +75,82 @@ Tensor cosine_similarity(const Tensor& x1, const Tensor& x2, int64_t dim, double
   return w12.div_((w1 * w2).clamp_min_(eps));
 }
 
+Tensor to_squareform(const Tensor& self) 
+{
+  AT_CHECK(self.dim() == 1,
+    "expected vector-form distance as input, dim as 1, but got ", self.dim());
```
Comment (gchanan) [inline aten/src/ATen/native/Distance.cpp]: this error message is a little awkward.  How about something like "Expected 1-dimensional (vector-form) distance, but got tensor of dimensionality "...

File: aten/src/ATen/native/Distance.cpp [hunk 4]
```diff
@@ -75,4 +75,82 @@ Tensor cosine_similarity(const Tensor& x1, const Tensor& x2, int64_t dim, double
...
+}
+
+Tensor from_squareform(const Tensor& self) 
+{
+  AT_CHECK(self.dim() == 2,
+    "expected square-form distance as input, dim as 2, but got ", self.dim());
+  AT_CHECK(self.size(0) == self.size(1),
+    "expected square-form distance as input, size(0) ",
+    self.size(0), " != size(1) ", self.size(1));
+  int64_t d = self.size(0), size = d * (d - 1) / 2;
```
Comment (gchanan) [inline aten/src/ATen/native/Distance.cpp]: can you put these on separate lines?

File: aten/src/ATen/native/Distance.cpp [hunk 5]
```diff
@@ -75,4 +75,82 @@ Tensor cosine_similarity(const Tensor& x1, const Tensor& x2, int64_t dim, double
   return w12.div_((w1 * w2).clamp_min_(eps));
 }
 
+Tensor to_squareform(const Tensor& self) 
+{
+  AT_CHECK(self.dim() == 1,
+    "expected vector-form distance as input, dim as 1, but got ", self.dim());
+  int64_t size = self.size(0), d = (std::sqrt(8 * size + 1) + 1) / 2;
```
Comment (gchanan) [inline aten/src/ATen/native/Distance.cpp]: nit: can you put these on separate lines?

File: aten/src/ATen/native/Distance.cpp [hunk 6]
```diff
@@ -75,4 +75,82 @@ Tensor cosine_similarity(const Tensor& x1, const Tensor& x2, int64_t dim, double
...
+  AT_CHECK(d * (d - 1) / 2 == size,
+    "not able to derive dimension, d from size ", size);
+  Tensor out_tensor = at::zeros({d, d}, self.options());
+  for (int64_t i = 0, idx = 0; i < d; i++)
+  {
+    for (int64_t j = i + 1; j < d && idx < size; j++)
+    {
+      out_tensor[i][j] = self[idx];
```
Comment (gchanan) [inline aten/src/ATen/native/Distance.cpp]: I think this will be slow because you create multiple tensors each time through the loop (can you benchmark against numpy?).  Instead you could operate on .data and use the strides.

File: aten/src/ATen/native/Distance.cpp (hunk 5 above)
Comment (gchanan) [reply aten/src/ATen/native/Distance.cpp]: can you add a comment explaining the "d" calculation?

**Comment by fmassa** [comment]
Question: why not dispatching to the `index` / `index_put_` kernels with the right indexing elements? This would involve some extra iterations maybe (generating the indices), but it would support GPU and backward computation off the box

**Comment by ezyang** [comment]
@fmassa I suggested to @lhuang04 to implement it directly, mostly because it didn't seem wise to have to generate another tensor (order size of the output) to setup the indices, when it's completely avoidable. Also `index_put_` won't be differentiable ;)

**Comment by lhuang04** [comment]
so far the benchmark status is like this,
#scipy squareform
>>> timeit("foo(lst, squareform)", globals=globals(), number=100)
0.02014931500889361
#current CPU implementation
>>> timeit("foo(lst, torch.squareform)", globals=globals(), number=100)
0.01594047201797366
#current GPU implementation(still in progress)
>>> timeit("foo(clst, torch.squareform)", globals=globals(), number=100)
2.789320666110143

File: aten/src/ATen/native/Distance.cpp [hunk 7]
```diff
@@ -75,4 +88,269 @@ Tensor cosine_similarity(const Tensor& x1, const Tensor& x2, int64_t dim, double
...
+      }
+      );
+  return out_tensor;
+}
+
+Tensor squareform_cpu(const Tensor& self)
+{
+  AT_CHECK((self.dim() == 1 || self.dim() == 2),
+    "expected vector-form distance or square-form distance as input",
+    " but the dimension got from input is ", self.dim());
```
Comment (gchanan) [inline aten/src/ATen/native/Distance.cpp]: nit: dimension of input is.

File: aten/src/ATen/native/Distance.cpp [hunk 8]
```diff
@@ -75,4 +88,269 @@ Tensor cosine_similarity(const Tensor& x1, const Tensor& x2, int64_t dim, double
...
+    "Expected 1-dimensional (vector-form) distance,",
+    " but got tensor of dimensionality ", self.dim());
+  int64_t size = self.size(0);
+  int64_t d = getSquareFormDimension(size);
+  Tensor out_tensor = at::zeros({d, d}, self.options());
+  AT_DISPATCH_ALL_TYPES(self.type(), "squareform", [&] {
+      to_squareform_frame<scalar_t> (self.data<scalar_t>(),
+        out_tensor.data<scalar_t>(), d, size);
+      }
+      );
```
Comment (gchanan) [inline aten/src/ATen/native/Distance.cpp]: nit: spacing is wrong here.

File: aten/src/ATen/native/Distance.cpp [hunk 9]
```diff
@@ -75,4 +88,269 @@ Tensor cosine_similarity(const Tensor& x1, const Tensor& x2, int64_t dim, double
   return w12.div_((w1 * w2).clamp_min_(eps));
 }
 
+template <typename scalar_t>
+static void to_squareform_frame(
+    scalar_t *input_p, scalar_t *output_p, int64_t dimension, int64_t size)
+{
+  for (int64_t i = 0, idx = 0; i < dimension; i++)
```
Comment (gchanan) [inline aten/src/ATen/native/Distance.cpp]: this won't work if the input isn't contiguous (add a test).

File: aten/src/ATen/native/Distance.cpp [hunk 10]
```diff
@@ -75,4 +88,269 @@ Tensor cosine_similarity(const Tensor& x1, const Tensor& x2, int64_t dim, double
   return w12.div_((w1 * w2).clamp_min_(eps));
 }
 
+template <typename scalar_t>
+static void to_squareform_frame(
+    scalar_t *input_p, scalar_t *output_p, int64_t dimension, int64_t size)
```
Comment (gchanan) [inline aten/src/ATen/native/Distance.cpp]: you are using terms in non-standard ways which makes this a little difficult to read.  dimension usually refers to an dimension index and size usually refers to the size of the dimension, but these are both referring to sizes here.  How about something like "square_dim_size" and "vector_size" or something?

File: aten/src/ATen/native/Distance.cpp [hunk 11]
```diff
@@ -75,4 +88,269 @@ Tensor cosine_similarity(const Tensor& x1, const Tensor& x2, int64_t dim, double
...
+{
+  for (int64_t i = 0, idx = 0; i < dimension; i++)
+  {
+    for (int64_t j = i + 1; j < dimension && idx < size; j++)
```
Comment (gchanan) [inline aten/src/ATen/native/Distance.cpp]: do you need the idx < size?

File: aten/src/ATen/native/Distance.cpp [hunk 12]
```diff
@@ -75,4 +88,269 @@ Tensor cosine_similarity(const Tensor& x1, const Tensor& x2, int64_t dim, double
   return w12.div_((w1 * w2).clamp_min_(eps));
 }
 
+template <typename scalar_t>
+static void to_squareform_frame(
```
Comment (gchanan) [inline aten/src/ATen/native/Distance.cpp]: can you put all the to_squareform and from_squareform things together, respectively?  This is unnecessarily difficult to read.

File: aten/src/ATen/native/Distance.cpp [hunk 13]
```diff
@@ -75,4 +88,269 @@ Tensor cosine_similarity(const Tensor& x1, const Tensor& x2, int64_t dim, double
...
+  return out_tensor;
+}
+
+Tensor from_squareform_cpu(const Tensor& self)
+{
+  AT_CHECK(self.dim() == 2,
+    "Expected 2-dimensional (square-form) distance,",
+    " but got tensor of dimensionality ", self.dim());
+  AT_CHECK(self.size(0) == self.size(1),
+    "expected square-form distance as input, size(0) ",
```
Comment (gchanan) [inline aten/src/ATen/native/Distance.cpp]: this isn't a particularly helpful error message; the user might not know what a square-form distance is.  What about just saying square matrix?

File: aten/src/ATen/native/Distance.cpp [hunk 14]
```diff
@@ -75,4 +88,269 @@ Tensor cosine_similarity(const Tensor& x1, const Tensor& x2, int64_t dim, double
...
+}
+
+Tensor to_squareform_cpu(const Tensor& self)
+{
+  AT_CHECK(self.dim() == 1,
+    "Expected 1-dimensional (vector-form) distance,",
+    " but got tensor of dimensionality ", self.dim());
+  int64_t size = self.size(0);
+  int64_t d = getSquareFormDimension(size);
+  Tensor out_tensor = at::zeros({d, d}, self.options());
```
Comment (gchanan) [inline aten/src/ATen/native/Distance.cpp]: it's wasteful to set all the values to 0 when you only need the diagonals to be 0.

File: aten/src/ATen/native/Distance.cpp [hunk 15]
```diff
@@ -75,4 +88,269 @@ Tensor cosine_similarity(const Tensor& x1, const Tensor& x2, int64_t dim, double
...
+      );
+  return out_tensor;
+}
+
+Tensor to_squareform_backward_cpu(const Tensor& grad, const Tensor& self)
+{
+  AT_CHECK(grad.dim() == 2,
+    "Expected 2D grad, but the dimension of the grad is ", grad.dim());
+  AT_CHECK(grad.size(0) == grad.size(1),
+    "expected square-form distance as input, size(0) ",
```
Comment (gchanan) [inline aten/src/ATen/native/Distance.cpp]: same here.

File: aten/src/ATen/native/Distance.cpp [hunk 16]
```diff
@@ -75,4 +88,269 @@ Tensor cosine_similarity(const Tensor& x1, const Tensor& x2, int64_t dim, double
...
+    grad.size(0), " != size(1) ", grad.size(1));
+  int64_t d = grad.size(0);
+  int64_t size = d * (d - 1) / 2;
+  Tensor out_tensor = at::zeros({size}, grad.options());
+  AT_DISPATCH_ALL_TYPES(self.type(), "squareform", [&] {
+      to_squareform_backward_frame<scalar_t> (grad.data<scalar_t>(),
+        out_tensor.data<scalar_t>(), d, size);
+      }
+      );
```
Comment (gchanan) [inline aten/src/ATen/native/Distance.cpp]: spacing.

File: test/test_torch.py [hunk 17]
```diff
@@ -1140,6 +1140,29 @@ def test_pdist_norm(self):
...
+            self.assertEqual(vec, torch.squareform(torch.squareform(vec)))
+            # grad check
+            vec = torch.arange(1, 7, dtype=torch.double, requires_grad=True, device=device)
+            full_square = torch.tensor(
+              [[0, 1, 2, 3], [1, 0, 4, 5], [2, 4, 0, 6], [3, 5, 6, 0]],
+              device=device,
+              dtype=torch.double,
+              requires_grad=True
+            )
+            self.assertTrue(torch.autograd.gradcheck(lambda x: torch.squareform(x), (vec)))
```
Comment (gchanan) [inline test/test_torch.py]: don't do gradchecks in here, do it in test_autograd.

File: aten/src/ATen/native/Distance.cpp [hunk 18]
```diff
@@ -75,4 +88,269 @@ Tensor cosine_similarity(const Tensor& x1, const Tensor& x2, int64_t dim, double
...
+      out_tensor[idx] = self[i][j];
+      idx++;
+    }
+  }
+  return out_tensor;
+}
+
+Tensor to_squareform_backward_cuda(const Tensor& grad, const Tensor& self)
+{
+  AT_CHECK(grad.dim() == 2,
```
Comment (gchanan) [inline aten/src/ATen/native/Distance.cpp]: this cuda stuff shouldn't be here, this isn't a cuda implementation.  just check if the input is cuda and throw an error (or put in a dispatch only for CPU).

File: test/test_torch.py [hunk 19]
```diff
@@ -1140,6 +1140,29 @@ def test_pdist_norm(self):
                         self.assertEqual(expected.shape, actual.shape)
                         self.assertTrue(torch.allclose(expected, actual))
 
+    def test_squareform(self):
+        devices = ['cpu'] if not torch.cuda.is_available() else ['cpu', 'cuda']
+        for device in devices:
+            vec = torch.arange(1, 7, device=device)
```
Comment (gchanan) [inline test/test_torch.py]: how about edge cases, i.e. d == 0, d == 1, (0,0), (1,1), error cases?

File: aten/src/ATen/native/SquareForm.cpp [hunk 20]
```diff
@@ -0,0 +1,181 @@
...
+    "]. It must be a binomial coefficient n choose 2 for some integer");
+  return d;
+}
+}
+
+template <typename scalar_t>
+static void to_squareform_frame(
+    scalar_t *input_p, scalar_t *output_p, int64_t square_dim_size, int64_t vector_size)
+{
+  for (int64_t i = 0, idx = 0; i < square_dim_size; i++)
```
Comment (gchanan) [inline aten/src/ATen/native/SquareForm.cpp]: you should write a cuda kernel, see the other examples in this directory.

**Comment by lhuang04** [comment]
# Benchmark in ipython
import torch
from scipy.spatial.distance import squareform, pdist
cuda = torch.device('cuda')
#1024x1024
v523776 = torch.arange(1, 523777, dtype=torch.double)
#2048x2048
v2096128 = torch.arange(1, 2096129, dtype=torch.double)
#10000x10000
v49995000 = torch.arange(1, 49995001, dtype=torch.double)

c523776 = torch.arange(1, 523777, dtype=torch.double, device = cuda)
c2096128 = torch.arange(1, 2096129, dtype=torch.double, device = cuda)
c49995000 = torch.arange(1, 49995001, dtype=torch.double, device = cuda)
lst = [v523776, v2096128, v49995000]
clst = [c523776, c2096128, c49995000]

def foo(lst, f):
  for vec in lst:
    ysq = f(vec)
    yvec = f(ysq)


timeit foo(lst, squareform)

timeit foo(lst, torch.squareform)

timeit foo(clst, torch.squareform)

# scipy
In [5]: timeit foo(lst, squareform)
3.62 s ± 28.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
# torch cpu
In [6]: timeit foo(lst, torch.squareform)
2.11 s ± 49.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
# torch cuda
In [7]: timeit foo(clst, torch.squareform)
41.6 ms ± 11.3 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

File: aten/src/ATen/native/SquareForm.cpp [hunk 21]
```diff
@@ -0,0 +1,168 @@
+#include "ATen/ATen.h"
```
Comment (gchanan) [inline aten/src/ATen/native/SquareForm.cpp]: use <> instead of ""

File: test/test_autograd.py [hunk 22]
```diff
@@ -2831,21 +2831,23 @@ def check(name):
...
-                    output = getattr(inputs[0], name)(*inputs[1:], **kwargs)
-                    return output_process_fn(output)
-
-                if not is_inplace and name not in EXCLUDE_GRADCHECK:
-                    run_grad_and_gradgrad_checks(self, name, test_name, fn,
-                                                 output_variable, (self_variable,) + args_variable)
+                is_method = hasattr(self_variable, name)
+                if is_method:
+                    output_variable = getattr(self_variable, name)(*args_variable, **kwargs_variable)
+                    if not exclude_tensor_method(name, test_name):
```
Comment (gchanan) [inline test/test_autograd.py]: is exclude_tensor_method here not sufficient?

File: test/test_torch.py [hunk 23]
```diff
@@ -1140,6 +1140,45 @@ def test_pdist_norm(self):
...
+            full_square_ret = torch.squareform(vec)
+            self.assertEqual(full_square_ret, full_square, "{} did not equal expected result".format(full_square_ret))
+            vec_ret = torch.squareform(full_square)
+            self.assertEqual(vec_ret, vec, "{} did not equal expected result".format(vec))
+            # test symmetric
+            self.assertEqual(full_square, torch.squareform(torch.squareform(full_square)))
+            self.assertEqual(vec, torch.squareform(torch.squareform(vec)))
+
+            # test contiguous
+            non_contiguous_vec = torch.tensor([[0, 1], [0, 2], [0, 3], [0, 4], [0, 5], [0, 6]])[:, 1]
```
Comment (gchanan) [inline test/test_torch.py]: assert it isn't contiguous.

File: test/test_torch.py [hunk 24]
```diff
@@ -1140,6 +1140,45 @@ def test_pdist_norm(self):
...
+            full_square_ret = torch.squareform(non_contiguous_vec)
+            self.assertEqual(full_square_ret, full_square, "{} did not equal expected result".format(full_square_ret))
+
+            # expected failure case
+            vec = torch.arange(1, 58, device=device)
+            self.assertRaises(RuntimeError, lambda: torch.squareform(vec))
```
Comment (gchanan) [inline test/test_torch.py]: can you check the message?

File: test/test_autograd.py (hunk 22 above)
Comment (lhuang04) [reply test/test_autograd.py]: no, it is not sufficient, the line above it (2834) will fail if the name is not a tensor method.

Comment (lhuang04) [reply test/test_autograd.py]: 2836 in the new code

Comment (gchanan) [reply test/test_autograd.py]: FYI you can link the code directly, click "view file" under "Files Changed" such as 
`https://github.com/pytorch/pytorch/blob/c85170bd2d0bc09c4bc91c621b96345b01069f16/test/test_autograd.py#L2836`

which is displayed as:
https://github.com/pytorch/pytorch/blob/c85170bd2d0bc09c4bc91c621b96345b01069f16/test/test_autograd.py#L2836.

File: test/common_methods_invocations.py [hunk 25]
```diff
@@ -1023,6 +1031,10 @@ def unpack_variables(args):
     'test_slogdet_1x1_neg_det',
     'test_slogdet_neg_det',
     'test_slogdet_symmetric',
+    'test_squareform_1d',
```
Comment (gchanan) [inline test/common_methods_invocations.py]: why are these included here?

File: test/common_methods_invocations.py [hunk 26]
```diff
@@ -797,6 +797,14 @@ def method_tests():
     ]
 # TODO: clamp with min/max
 
+def torch_function_tests():
+    return [
+        ('squareform', (1,), NO_ARGS, '1d'),
```
Comment (gchanan) [inline test/common_methods_invocations.py]: the point of this file is to automatically run these tests through different functionality, i.e. autograd, jit, etc.  This won't be run through the JIT tests -- does the JIT work if you add these tests to the above list?

File: test/test_autograd.py [hunk 27]
```diff
@@ -2831,21 +2831,23 @@ def check(name):
...
+                            output_tensor = torch.DoubleTensor((output_tensor,))
+                        self.assertEqual(unpack_variables(output_variable), output_tensor)
+                        # TODO: check that both have changed after adding all inplace ops
+
+                    def fn(*inputs):
+                        output = getattr(inputs[0], name)(*inputs[1:], **kwargs)
+                        return output_process_fn(output)
+
+                    if not is_inplace and name not in EXCLUDE_GRADCHECK:
+                        run_grad_and_gradgrad_checks(self, name, test_name, fn,
```
Comment (gchanan) [inline test/test_autograd.py]: I don't really get what the existing tests are doing that exclude_tensor_method but still go here.  Can you explain?

File: test/test_torch.py [hunk 28]
```diff
@@ -1139,6 +1139,57 @@ def test_pdist_norm(self):
...
 
+    def test_squareform(self):
+        devices = ['cpu'] if not torch.cuda.is_available() else ['cpu', 'cuda']
+        for device in devices:
+            # smallest case []
+            vec = torch.empty(0, device=device)
+            sq = [[0]]
+            full_square = torch.tensor(sq, device=device)
+            full_square_ret = torch.squareform(vec)
+            self.assertEqual(full_square_ret, full_square, "{} did not equal expected result".format(full_square_ret))
```
Comment (gchanan) [inline test/test_torch.py]: shouldn't the default message be something like you added here?

File: test/test_torch.py [hunk 29]
```diff
@@ -1139,6 +1139,57 @@ def test_pdist_norm(self):
...
+
+            # expected failure case
+            vec = torch.arange(1, 58, device=device)
+            self.assertRaises(RuntimeError, lambda: torch.squareform(vec))
+            rect = torch.zeros((4, 2), device=device)
+            self.assertRaises(RuntimeError, lambda: torch.squareform(rect))
+            # test error message
+            try:
+                torch.squareform(vec)
```
Comment (gchanan) [inline test/test_torch.py]: I think assertRaisesRegex is what you want.

File: test/test_torch.py [hunk 30]
```diff
@@ -1139,6 +1139,57 @@ def test_pdist_norm(self):
...
+            vec = torch.empty(0, device=device)
+            sq = [[0]]
+            full_square = torch.tensor(sq, device=device)
+            full_square_ret = torch.squareform(vec)
+            self.assertEqual(full_square_ret, full_square, "{} did not equal expected result".format(full_square_ret))
+
+            # smallest case [1]
+            vec = torch.arange(1, 2, device=device)
+            sq = [[0, 1], [1, 0]]
+            full_square = torch.tensor(sq, device=device)
```
Comment (gchanan) [inline test/test_torch.py]: what about tests for valid sizes, but invalid inputs, e.g. non-symmetric matrices?

**Comment by LemonPi** [comment]
This would be nice to have for the current torch.pdist function; is this still being worked on?

**Comment by TobiasFischerP4D** [comment]
I am wondering the same thing. Is this under active development? Looks like it is quite close to being finished!

**Comment by ezyang** [comment]
Nope, this diff has been abandoned. If someone wants to pick it up and drive it through, that would be great!
